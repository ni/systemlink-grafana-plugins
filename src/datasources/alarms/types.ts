import { DataQuery } from '@grafana/schema';

export interface AlarmsQuery extends DataQuery {}
