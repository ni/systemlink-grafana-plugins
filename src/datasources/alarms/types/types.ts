import { DataQuery } from '@grafana/schema';

export interface AlarmsQuery extends DataQuery {
  queryType?: QueryType;
}

export interface AlarmsVariableQuery extends AlarmsQuery {
  filter?: string;
  descending?: boolean;
  take?: number;
}

export enum QueryType {
  AlarmsCount = 'Alarms Count',
  ListAlarms = 'List Alarms'
}

export interface QueryAlarmsRequest {
  filter?: string;
  transitionInclusionOption?: TransitionInclusionOption;
  take?: number;
  orderBy?: string;
  orderByDescending?: boolean;
  continuationToken?: string;
  returnCount?: boolean;
  returnMostRecentlyOccurredOnly?: boolean;
}

export interface QueryAlarmsResponse {
  alarms: Alarm[];
  totalCount?: number;
  continuationToken?: string;
}

export interface Alarm {
  instanceId: string;
  alarmId: string;
  workspace: string;
  active: boolean;
  clear: boolean;
  acknowledged: boolean;
  acknowledgedAt: string | null;
  acknowledgedBy: string | null;
  occurredAt: string;
  updatedAt: string;
  createdBy: string;
  transitions: AlarmTransition[];
  transitionOverflowCount: number;
  currentSeverityLevel: number;
  highestSeverityLevel: number;
  mostRecentSetOccurredAt: string | null;
  mostRecentTransitionOccurredAt: string | null;
  channel: string;
  condition: string;
  displayName: string;
  description: string;
  keywords: string[];
  properties: {
    [key: string]: string;
  };
  resourceType: string;
}

enum TransitionInclusionOption {
  None = 'NONE',
  MostRecentOnly = 'MOST_RECENT_ONLY',
  All = 'ALL',
};

interface AlarmTransition {
  transitionType: AlarmTransitionType;
  occurredAt: string;
  severityLevel: number;
  value: string;
  condition: string;
  shortText: string;
  detailText: string;
  keywords: string[];
  properties: {
    [key: string]: string;
  };
}

export enum AlarmTransitionType {
  Clear = 'CLEAR',
  Set = 'SET',
};
