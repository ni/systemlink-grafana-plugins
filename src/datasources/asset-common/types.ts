
export interface AssetsResponse {
    assets: AssetModel[],
    totalCount: number
  }
  
  export interface AssetLocationModel {
    minionId: string,
    parent: string,
    resourceUri: string,
    slotNumber: number,
    state: AssetPresenceWithSystemConnectionModel,
    physicalLocation?: string
  }
  
  export interface AssetPresenceWithSystemConnectionModel {
    assetPresence: "INITIALIZING" | "UNKNOWN" | "NOT_PRESENT" | "PRESENT" | string,
    // to be compatible with both SLS and SLE
    systemConnection?: "APPROVED" | "DISCONNECTED" | "CONNECTED_UPDATE_PENDING" | "CONNECTED_UPDATE_SUCCESSFUL" | "CONNECTED_UPDATE_FAILED" | "UNSUPPORTED" | "ACTIVATED" | "CONNECTED" | string
  }
  
  export interface AssetModel {
    modelName: string,
    modelNumber: number,
    serialNumber: string,
    vendorName: string,
    vendorNumber: number,
    busType: 'BUILT_IN_SYSTEM' | 'PCI_PXI' | 'USB' | 'GPIB' | 'VXI' | 'SERIAL' | 'TCP_IP' | 'CRIO' | 'SCXI' | 'CDAQ' | 'SWITCH_BLOCK' | 'SCC' | 'FIRE_WIRE' | 'ACCESSORY' | 'CAN' | 'SWITCH_BLOCK_DEVICE' | 'SLSC' | string,
    name: string,
    assetType: 'GENERIC' | 'DEVICE_UNDER_TEST' | 'FIXTURE' | 'SYSTEM' | string,
    firmwareVersion: string,
    hardwareVersion: string,
    visaResourceName: string,
    temperatureSensors: any[],
    supportsSelfCalibration: boolean,
    supportsExternalCalibration: boolean,
    customCalibrationInterval?: number,
    selfCalibration?: any,
    isNIAsset: boolean,
    workspace: string,
    fileIds: string[],
    supportsSelfTest: boolean,
    supportsReset: boolean,
    id: string,
    location: AssetLocationModel,
    calibrationStatus: 'OK' | 'APPROACHING_RECOMMENDED_DUE_DATE' | 'PAST_RECOMMENDED_DUE_DATE' | string,
    isSystemController: boolean,
    externalCalibration?: ExternalCalibrationModel,
    discoveryType: 'MANUAL' | 'AUTOMATIC' | string,
    properties: Record<string, string>,
    keywords: string[],
    lastUpdatedTimestamp: string,
  }
  
  export interface ExternalCalibrationModel {
    temperatureSensors: any[],
    isLimited?: boolean,
    date: string,
    recommendedInterval: number,
    nextRecommendedDate: string,
    nextCustomDueDate?: string,
    comments: string,
    entryType: "AUTOMATIC" | "MANUAL" | string,
    operator: ExternalCalibrationOperatorModel
  }
  
  export interface ExternalCalibrationOperatorModel {
    displayName: string;
    userId: string
  }
