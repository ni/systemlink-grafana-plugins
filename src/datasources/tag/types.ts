import { DataQuery } from '@grafana/schema';

export enum TagQueryType {
  Current = 'Current',
  History = 'History',
}

export interface TagQuery extends DataQuery {
  type: TagQueryType;
  path: string;
  workspace: string;
}

export interface TagWithValue {
  current: { value: { value: string } };
  tag: {
    path: string;
    properties: { displayName?: string } | null;
    workspace_id: string;
  };
}

export interface TagsWithValues {
  tagsWithValues: TagWithValue[];
}

export interface TagHistoryResponse {
  results: {
    [path: string]: {
      type: string;
      values: Array<{ timestamp: string; value: string }>;
    };
  };
}
