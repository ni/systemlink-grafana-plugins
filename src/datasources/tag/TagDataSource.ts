import {
  DataFrameDTO,
  DataSourceInstanceSettings,
  DataQueryRequest,
  TimeRange,
  FieldType,
  TestDataSourceResponse,
  FieldConfig,
  dateTime,
} from '@grafana/data';
import { BackendSrv, TemplateSrv, getBackendSrv, getTemplateSrv } from '@grafana/runtime';
import { DataSourceBase } from 'core/DataSourceBase';
import {
  TagHistoryResponse,
  TagQuery,
  TagQueryType,
  TagsWithValues,
  TagWithValue,
  TimeAndTagTypeValues,
  TypeAndValues,
} from './types';
import { Throw, getWorkspaceName } from 'core/utils';

export class TagDataSource extends DataSourceBase<TagQuery> {
  constructor(
    readonly instanceSettings: DataSourceInstanceSettings,
    readonly backendSrv: BackendSrv = getBackendSrv(),
    readonly templateSrv: TemplateSrv = getTemplateSrv()
  ) {
    super(instanceSettings, backendSrv, templateSrv);
  }

  tagUrl = this.instanceSettings.url + '/nitag/v2';
  tagHistoryUrl = this.instanceSettings.url + '/nitaghistorian/v2/tags';

  defaultQuery: Omit<TagQuery, 'refId'> = {
    type: TagQueryType.Current,
    path: '',
    workspace: '',
    properties: false,
  };

  async runQuery(query: TagQuery, { range, maxDataPoints, scopedVars }: DataQueryRequest): Promise<DataFrameDTO> {
    const tagsWithValues = await this.getMostRecentTags(
      this.templateSrv.replace(query.path, scopedVars),
      this.templateSrv.replace(query.workspace, scopedVars)
    );
    const workspaces = await this.getWorkspaces();
    const result: DataFrameDTO = { refId: query.refId, fields: [] };

    if (query.type === TagQueryType.Current) {
      const allPossibleProps = this.getAllProperties(tagsWithValues);
      result.fields = [
        {
          name: 'name',
          values: tagsWithValues.map((tag: TagWithValue) => tag.tag.properties?.displayName || tag.tag.path)
        },
        {
          name: 'value',
          values: tagsWithValues.map((tag: TagWithValue) => this.convertTagValue(tag.tag.type ?? tag.tag.datatype, tag.current?.value.value)),
        },
        {
          name: 'updated',
          values: tagsWithValues.map((tag: TagWithValue) => tag.current?.timestamp),
          type: FieldType.time,
          config: { unit: 'dateTimeFromNow' }
        }
      ]
      if (query.properties) {
        allPossibleProps.forEach((prop) => {
          result.fields.push(
            {
              name: prop,
              values: tagsWithValues.map((tag: TagWithValue) => tag.tag.properties && tag.tag.properties[prop] ? tag.tag.properties[prop] : '')
            }
          );
        });
      }

      return result
    } else {
      const workspaceTagMap: Record<string, TagWithValue[]> = {};
      const tagPropertiesMap: Record<string, Record<string, string> | null> = {};

      // Identify tags that exist in more than one workspace
      const tagPathCount: Record<string, number> = {};
      for (const tagWithValue of tagsWithValues) {
        tagPathCount[tagWithValue.tag.path] = (tagPathCount[tagWithValue.tag.path] || 0) + 1;
      }

      for (const tagWithValue of tagsWithValues) {
        const workspace = tagWithValue.tag.workspace ?? tagWithValue.tag.workspace_id;
        if (!workspaceTagMap[workspace]) {
          workspaceTagMap[workspace] = [];
        }
        workspaceTagMap[workspace].push(tagWithValue);
        const prefixedPath = tagPathCount[tagWithValue.tag.path] > 1
          ? `${getWorkspaceName(workspaces, workspace)}.${tagWithValue.tag.path}`
          : tagWithValue.tag.path;
        tagPropertiesMap[prefixedPath] = tagWithValue.tag.properties;
      }

      let tagsDecimatedHistory: { [key: string]: TypeAndValues } = {};
      for (const workspace in workspaceTagMap) {
        const tagHistoryResponse = await this.getTagHistoryWithChunks(
          workspaceTagMap[workspace],
          workspace,
          range,
          maxDataPoints
        )
        for (const path in tagHistoryResponse.results) {
          const prefixedPath = tagPathCount[path] > 1
            ? `${getWorkspaceName(workspaces, workspace)}.${path}`
            : path;
          tagsDecimatedHistory[prefixedPath] = tagHistoryResponse.results[path];
        }
      }

      const mergedTagValuesWithType = this.mergeTagsHistoryValues(tagsDecimatedHistory);
      result.fields.push({
        name: 'time', values: mergedTagValuesWithType.timestamps.map(v => dateTime(v).valueOf()), type: FieldType.time
      });

      for (const path in mergedTagValuesWithType.values) {
        const config: FieldConfig = {};
        const tagProps = tagPropertiesMap[path]
        if (tagProps?.units) {
          config.unit = tagProps.units
        }
        if (tagProps?.displayName) {
          config.displayName = tagProps.displayName
          config.displayNameFromDS = tagProps.displayName
        }
        result.fields.push({
          name: path,
          values: mergedTagValuesWithType.values[path].values.map((value) => {
            return this.convertTagValue(mergedTagValuesWithType.values[path].type, value)
          }),
          config
        });
      }

      return result
    }
  }

  private async getMostRecentTags(path: string, workspace: string) {
    let filter = `path = "${path}"`;
    if (workspace) {
      filter += ` && workspace = "${workspace}"`;
    }

    const response = await this.post<TagsWithValues>(this.tagUrl + '/query-tags-with-values', {
      filter,
      take: 32,
      orderBy: 'TIMESTAMP',
      descending: true,
    });

    return response.tagsWithValues.length ? response.tagsWithValues : Throw(`No tags matched the path '${path}'`)
  }

  private async getTagHistoryWithChunks(paths: TagWithValue[], workspace: string, range: TimeRange, intervals?: number): Promise<TagHistoryResponse> {
    const pathChunks: TagWithValue[][] = [];
    for (let i = 0; i < paths.length; i += 10) {
      pathChunks.push(paths.slice(i, i + 10));
    }
    // Fetch and aggregate the data from each chunk
    const aggregatedResults: TagHistoryResponse = { results: {} };
    for (const chunk of pathChunks) {
      const chunkResult = await this.getTagHistoryValues(chunk.map(tag => tag.tag.path), workspace, range, intervals);
      // Merge the results from the current chunk with the aggregated results
      for (const [path, data] of Object.entries(chunkResult.results)) {
        if (!aggregatedResults.results[path]) {
          aggregatedResults.results[path] = data;
        } else {
          aggregatedResults.results[path].values = aggregatedResults.results[path].values.concat(data.values);
        }
      }
    }

    return aggregatedResults;
  }

  private async getTagHistoryValues(paths: string[], workspace: string, range: TimeRange, intervals?: number): Promise<TagHistoryResponse> {
    return await this.post<TagHistoryResponse>(`${this.tagHistoryUrl}/query-decimated-history`, {
      paths,
      workspace,
      startTime: range.from.toISOString(),
      endTime: range.to.toISOString(),
      decimation: intervals ? Math.min(intervals, 1000) : 500,
    });
  };

  private convertTagValue(type: string, value?: string) {
    return value && ['DOUBLE', 'INT', 'U_INT64'].includes(type) ? Number(value) : value;
  }

  private getAllProperties(data: TagWithValue[]) {
    const props: Set<string> = new Set();
    data.forEach((tag) => {
      if (tag.tag.properties) {
        Object.keys(tag.tag.properties)
          .filter(name => !name.startsWith('nitag'))
          .forEach((name) => {
            props.add(name)
          })
      }
    })

    return props
  }

  shouldRunQuery(query: TagQuery): boolean {
    return Boolean(query.path);
  }

  mergeTagsHistoryValues = (history: Record<string, TypeAndValues>): TimeAndTagTypeValues => {
    const timestampsSet: Set<string> = new Set();
    const values: TimeAndTagTypeValues = {
      timestamps: [],
      values: {}
    };
    for (const path in history) {
      for (const { timestamp } of history[path].values) {
        timestampsSet.add(timestamp);
      }
    }
    // Uniq timestamps from history data
    const timestamps = [...timestampsSet];
    // Sort timestamps to ensure a consistent order
    timestamps.sort();
    values.timestamps = timestamps;

    // Initialize arrays for each key
    for (const path in history) {
      values.values[path] = { 'type': history[path].type, 'values': new Array(timestamps.length).fill(null) };
    }
    // Populate the values arrays
    for (const path in history) {
      for (const historicalValue of history[path].values) {
        const index = timestamps.indexOf(historicalValue.timestamp);
        values.values[path]['values'][index] = historicalValue.value;
      }
    }

    return values;
  }

  async testDatasource(): Promise<TestDataSourceResponse> {
    await this.get(this.tagUrl + '/tags-count');
    return { status: 'success', message: 'Data source connected and authentication successful!' };
  }
}
