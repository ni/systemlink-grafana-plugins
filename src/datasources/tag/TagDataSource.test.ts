import { MockProxy } from 'jest-mock-extended';
import { TagDataSource } from './TagDataSource';
import { setupDataSource, createFetchError, getQueryBuilder, defaultQueryOptions } from 'test/fixtures';
import { BackendSrv, TemplateSrv } from '@grafana/runtime';
import { TagHistoryResponse, TagQuery, TagQueryType, TagsWithValues } from './types';

let ds: TagDataSource, backendSrv: MockProxy<BackendSrv>, templateSrv: MockProxy<TemplateSrv>;

beforeEach(() => {
  [ds, backendSrv, templateSrv] = setupDataSource(TagDataSource);
});

const buildQuery = getQueryBuilder<TagQuery>()({ type: TagQueryType.Current, workspace: '' });

describe('testDatasource', () => {
  test('returns success', async () => {
    backendSrv.get.calledWith('/nitag/v2/tags-count').mockResolvedValue(25);

    const result = await ds.testDatasource();

    expect(result.status).toEqual('success');
  });

  test('bubbles up exception', async () => {
    backendSrv.get.calledWith('/nitag/v2/tags-count').mockRejectedValue(createFetchError(400));

    await expect(ds.testDatasource()).rejects.toHaveProperty('status', 400);
  });
});

describe('queries', () => {
  test('tag current value', async () => {
    backendSrv.post
      .calledWith('/nitag/v2/query-tags-with-values', expect.objectContaining({ filter: 'path = "my.tag"' }))
      .mockResolvedValue(createQueryTagsResponse('my.tag', '3.14'));

    const result = await ds.query(buildQuery({ path: 'my.tag' }));

    expect(result.data).toEqual([
      {
        fields: [{ name: 'value', values: ['3.14'] }],
        name: 'my.tag',
        refId: 'A',
      },
    ]);
  });

  test('applies query defaults when missing fields', async () => {
    backendSrv.post.mockResolvedValue(createQueryTagsResponse('my.tag', '3.14'));

    const result = await ds.query({ ...defaultQueryOptions, targets: [{ path: 'my.tag'} as TagQuery]});

    expect(result.data[0]).toHaveProperty('fields', [{ name: 'value', values: ['3.14'] }]);
  });

  test('uses displayName property', async () => {
    backendSrv.post.mockResolvedValue(createQueryTagsResponse('my.tag', '3.14', 'My cool tag'));

    const result = await ds.query(buildQuery({ path: 'my.tag' }));

    expect(result.data[0]).toEqual(expect.objectContaining({ name: 'My cool tag' }));
  });

  test('multiple targets - skips invalid queries', async () => {
    backendSrv.post
      .mockResolvedValueOnce(createQueryTagsResponse('my.tag1', '3.14'))
      .mockResolvedValueOnce(createQueryTagsResponse('my.tag2', 'foo'));

    const result = await ds.query(buildQuery({ path: 'my.tag1' }, { path: '' }, { path: 'my.tag2' }));

    expect(backendSrv.post.mock.calls[0][1]).toHaveProperty('filter', 'path = "my.tag1"');
    expect(backendSrv.post.mock.calls[1][1]).toHaveProperty('filter', 'path = "my.tag2"');
    expect(result.data).toEqual([
      {
        fields: [{ name: 'value', values: ['3.14'] }],
        name: 'my.tag1',
        refId: 'A',
      },
      {
        fields: [{ name: 'value', values: ['foo'] }],
        name: 'my.tag2',
        refId: 'C',
      },
    ]);
  });

  test('throw when no tags matched', async () => {
    backendSrv.post.mockResolvedValue({ tagsWithValues: [] });

    await expect(ds.query(buildQuery({ path: 'my.tag' }))).rejects.toThrow(
      'my.tag'
    );
  });

  test('numeric tag history', async () => {
    const queryRequest = buildQuery({ type: TagQueryType.History, path: 'my.tag' });

    backendSrv.post
      .calledWith('/nitag/v2/query-tags-with-values', expect.objectContaining({ filter: 'path = "my.tag"' }))
      .mockResolvedValue(createQueryTagsResponse('my.tag', '3.14'));

    backendSrv.post
      .calledWith(
        '/nitaghistorian/v2/tags/query-decimated-history',
        expect.objectContaining({
          paths: ['my.tag'],
          workspace: '1',
          startTime: queryRequest.range.from.toISOString(),
          endTime: queryRequest.range.to.toISOString(),
          decimation: 300,
        })
      )
      .mockResolvedValue(
        createTagHistoryResponse('my.tag', 'DOUBLE', [
          { timestamp: '2023-01-01T00:00:00Z', value: '1' },
          { timestamp: '2023-01-01T00:01:00Z', value: '2' },
        ])
      );

    const result = await ds.query(queryRequest);

    expect(result.data).toEqual([
      {
        fields: [
          { name: 'time', values: [1672531200000, 1672531260000] },
          { name: 'value', values: [1, 2] },
        ],
        name: 'my.tag',
        refId: 'A',
      },
    ]);
  });

  test('string tag history', async () => {
    backendSrv.post.mockResolvedValueOnce(createQueryTagsResponse('my.tag', 'foo'));
    backendSrv.post.mockResolvedValueOnce(
      createTagHistoryResponse('my.tag', 'STRING', [
        { timestamp: '2023-01-01T00:00:00Z', value: '3.14' },
        { timestamp: '2023-01-01T00:01:00Z', value: 'foo' },
      ])
    );

    const result = await ds.query(buildQuery({ type: TagQueryType.History, path: 'my.tag' }));

    expect(result.data).toEqual([
      {
        fields: [
          { name: 'time', values: [1672531200000, 1672531260000] },
          { name: 'value', values: ['3.14', 'foo'] },
        ],
        name: 'my.tag',
        refId: 'A',
      },
    ]);
  });

  test('decimation parameter does not go above 1000', async () => {
    const queryRequest = buildQuery({ type: TagQueryType.History, path: 'my.tag' });
    queryRequest.maxDataPoints = 1500;

    backendSrv.post.mockResolvedValueOnce(createQueryTagsResponse('my.tag', '3'));

    backendSrv.post.mockResolvedValueOnce(
      createTagHistoryResponse('my.tag', 'INT', [
        { timestamp: '2023-01-01T00:00:00Z', value: '1' },
        { timestamp: '2023-01-01T00:01:00Z', value: '2' },
      ])
    );

    await ds.query(queryRequest);

    expect(backendSrv.post.mock.lastCall?.[1]).toHaveProperty('decimation', 1000);
  });

  test('replaces tag path with variable', async () => {
    templateSrv.replace.calledWith('$my_variable').mockReturnValue('my.tag');
    backendSrv.post
      .calledWith('/nitag/v2/query-tags-with-values', expect.objectContaining({ filter: 'path = "my.tag"' }))
      .mockResolvedValue(createQueryTagsResponse('my.tag', '3.14'));

    const result = await ds.query(buildQuery({ type: TagQueryType.Current, path: '$my_variable' }));

    expect(result.data[0]).toHaveProperty('name', 'my.tag');
  });

  test('filters by workspace if provided', async () => {
    backendSrv.post.mockResolvedValueOnce(createQueryTagsResponse('my.tag', '3.14', undefined, '2'));
    backendSrv.post.mockResolvedValueOnce(createTagHistoryResponse('my.tag', 'DOUBLE', []));

    await ds.query(buildQuery({ type: TagQueryType.History, path: 'my.tag', workspace: '2' }));

    expect(backendSrv.post.mock.calls[0][1]).toHaveProperty('filter', 'path = "my.tag" && workspace = "2"');
    expect(backendSrv.post.mock.calls[1][1]).toHaveProperty('workspace', '2');
  });
});

function createQueryTagsResponse(path: string, value: string, displayName?: string, workspace_id = '1'): TagsWithValues {
  return {
    tagsWithValues: [
      {
        current: { value: { value } },
        tag: { path, properties: { displayName }, workspace_id },
      },
    ],
  };
}

function createTagHistoryResponse(
  path: string,
  type: string,
  values: Array<{ timestamp: string; value: string }>
): TagHistoryResponse {
  return { results: { [path]: { type, values } } };
}
