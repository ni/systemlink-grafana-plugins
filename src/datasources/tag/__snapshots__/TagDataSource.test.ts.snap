// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`queries appends tag properties to query result 1`] = `
[
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "my.tag",
        ],
      },
      {
        "name": "value",
        "values": [
          3.14,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
      {
        "name": "foo",
        "values": [
          "bar",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries applies query defaults when missing fields 1`] = `
[
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "my.tag",
        ],
      },
      {
        "name": "value",
        "values": [
          3.14,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries current value for all data types 1`] = `
[
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "tag1",
        ],
      },
      {
        "name": "value",
        "values": [
          3,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
    ],
    "refId": "A",
  },
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "tag2",
        ],
      },
      {
        "name": "value",
        "values": [
          3.3,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
    ],
    "refId": "B",
  },
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "tag3",
        ],
      },
      {
        "name": "value",
        "values": [
          "foo",
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
    ],
    "refId": "C",
  },
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "tag4",
        ],
      },
      {
        "name": "value",
        "values": [
          "True",
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
    ],
    "refId": "D",
  },
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "tag5",
        ],
      },
      {
        "name": "value",
        "values": [
          2147483648,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
    ],
    "refId": "E",
  },
]
`;

exports[`queries handles null tag properties 1`] = `
[
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "my.tag",
        ],
      },
      {
        "name": "value",
        "values": [
          3.14,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries handles tag with no current value 1`] = `
[
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "my.tag",
        ],
      },
      {
        "name": "value",
        "values": [
          undefined,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          undefined,
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries multiple current values with properties 1`] = `
[
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "my.1.tag",
          "my.2.tag",
        ],
      },
      {
        "name": "value",
        "values": [
          3.14,
          41.3,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
          "2023-10-04T00:00:00.000000Z",
        ],
      },
      {
        "name": "unit",
        "values": [
          "A",
          "A",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries multiple current values with different properties 1`] = `
[
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "my.1.tag",
          "my.2.tag",
        ],
      },
      {
        "name": "value",
        "values": [
          3.14,
          41.3,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
          "2023-10-04T00:00:00.000000Z",
        ],
      },
      {
        "name": "upperCriticalThreshold",
        "values": [
          "15",
          "",
        ],
      },
      {
        "name": "unit",
        "values": [
          "",
          "A",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries multiple targets - skips invalid queries 1`] = `
[
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "my.tag1",
        ],
      },
      {
        "name": "value",
        "values": [
          3.14,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
    ],
    "refId": "A",
  },
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "my.tag2",
        ],
      },
      {
        "name": "value",
        "values": [
          41.3,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
    ],
    "refId": "C",
  },
]
`;

exports[`queries numeric tag history 1`] = `
[
  {
    "fields": [
      {
        "name": "time",
        "type": "time",
        "values": [
          1672531200000,
          1672531260000,
        ],
      },
      {
        "config": {},
        "name": "my.tag",
        "values": [
          1,
          2,
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries numeric multiple tags history 1`] = `
[
  {
    "fields": [
      {
        "name": "time",
        "type": "time",
        "values": [
          1672531200000,
          1672531260000,
          1672531320000,
        ],
      },
      {
        "config": {},
        "name": "my.tag.1",
        "values": [
          1,
          2,
          null,
        ],
      },
      {
        "config": {},
        "name": "my.tag.2",
        "values": [
          2,
          3,
          null,
        ],
      },
      {
        "config": {},
        "name": "my.tag.3",
        "values": [
          3,
          null,
          4,
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries replaces tag path with variable 1`] = `
[
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "my.tag",
        ],
      },
      {
        "name": "value",
        "values": [
          3.14,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries string tag history 1`] = `
[
  {
    "fields": [
      {
        "name": "time",
        "type": "time",
        "values": [
          1672531200000,
          1672531260000,
        ],
      },
      {
        "config": {},
        "name": "my.tag",
        "values": [
          "3.14",
          "foo",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries supports legacy tag service property "datatype" 1`] = `
[
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "my.tag",
        ],
      },
      {
        "name": "value",
        "values": [
          3.14,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries tag current value 1`] = `
[
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "my.tag",
        ],
      },
      {
        "name": "value",
        "values": [
          3.14,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries uses displayName property 1`] = `
[
  {
    "fields": [
      {
        "name": "name",
        "values": [
          "My cool tag",
        ],
      },
      {
        "name": "value",
        "values": [
          3.14,
        ],
      },
      {
        "config": {
          "unit": "dateTimeFromNow",
        },
        "name": "updated",
        "type": "time",
        "values": [
          "2023-10-04T00:00:00.000000Z",
        ],
      },
    ],
    "refId": "A",
  },
]
`;

exports[`queries add workspace prefix only when a tag with the same path exists in multiple workspaces 1`] = `
[
  {
    "fields": [
      {
        "name": "time",
        "type": "time",
        "values": [
          1672531200000,
          1672531260000,
        ],
      },
      {
        "config": {},
        "name": "Default workspace.my.tag.1",
        "values": [
          1,
          2,
        ],
      },
      {
        "config": {},
        "name": "Default workspace.my.tag.2",
        "values": [
          2,
          3,
        ],
      },
      {
        "config": {},
        "name": "Other workspace.my.tag.1",
        "values": [
          1,
          2,
        ],
      },
      {
        "config": {},
        "name": "Other workspace.my.tag.2",
        "values": [
          2,
          3,
        ],
      },
      {
        "config": {},
        "name": "my.tag.3",
        "values": [
          3,
          4,
        ],
      },
      {
        "config": {},
        "name": "my.tag.4",
        "values": [
          4,
          5,
        ],
      },
    ],
    "refId": "A",
  },
]
`;
